### Coded by Adel Samir ElZemity
### Udacity Data Analysis Professional Track
#### Date Created: 10 December 2020

In [2]:
# importing liberaries

import pandas as pd
import numpy as np
import requests
import json
import time
import matplotlib.pyplot as plt

# 1- Gathering the Data

In [3]:
# Reading Given Data
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
df_1 = pd.DataFrame(twitter_archive)

In [4]:
# Loading tweet images

with open('image_predictions.tsv' , 'wb') as file:
    image_predictions = requests.get('https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv', auth=('user', 'pass'))
    file.write(image_predictions.content)   

In [5]:
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

## Now let's test the data

In [6]:
twitter_archive.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [7]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [8]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [9]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [ ]:
# import tweepy
from tweepy import OAuthHandler
import json

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)
tweet_ids = df_1.tweet_id.values
len(tweet_ids)
# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API


# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}

# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass


print(fails_dict)


In [13]:
#reading tweets from json file
tweet_status = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')

In [14]:
tweet_status.head()

contributors  coordinates          created_at display_text_range  \
0           NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1           NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2           NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3           NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4           NaN          NaN 2017-07-29 16:00:24           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           39467   
1  {'media': [{'id': 892177413194625024, 'id_str'...           33819   
2  {'media': [{'id': 891815175371796480, 'id_str'...           25461   
3  {'media': [{'id': 891689552724799489, 'id_str'...           42908   
4  {'media': [{'id': 891327551943041024, 'id_str'...           41048   

   favorited                                          full_text  geo  \
0      False  This is Phineas. He's a mystical boy. Only eve...  NaN   
1      False  This is Tilly. She's just checking pup on you....  NaN   
2      False  This is Archie. He is a rare Norwegian Pouncin...  NaN   
3      False  This is Darla. She commenced a snooze mid meal...  NaN   
4      False  This is Franklin. He would like you to stop ca...  NaN   

                         ...                          \
0                        ...                           
1                        ...                           
2                        ...                           
3                        ...                           
4                        ...                           

   possibly_sensitive_appealable  quoted_status quoted_status_id  \
0                            0.0            NaN              NaN   
1                            0.0            NaN              NaN   
2                            0.0            NaN              NaN   
3                            0.0            NaN              NaN   
4                            0.0            NaN              NaN   

   quoted_status_id_str  retweet_count  retweeted  retweeted_status  \
0                   NaN           8853      False               NaN   
1                   NaN           6514      False               NaN   
2                   NaN           4328      False               NaN   
3                   NaN           8964      False               NaN   
4                   NaN           9774      False               NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 31 columns]

# Data Assessment

In [15]:
# Reading gathered files
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
tweet_status = pd.read_json('tweet_json.txt', lines = True,encoding='utf-8')

## Visual Assessment Part

In [41]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [44]:
image_predictions.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [46]:
tweet_status.sample(10)

contributors  coordinates          created_at display_text_range  \
313            NaN          NaN 2017-02-24 17:01:22           [0, 128]   
1266           NaN          NaN 2016-03-15 01:48:55           [0, 131]   
1942           NaN          NaN 2015-12-07 03:40:24           [0, 111]   
2109           NaN          NaN 2015-11-28 03:21:24           [0, 124]   
2028           NaN          NaN 2015-12-01 20:35:22           [0, 139]   
385            NaN          NaN 2017-02-01 02:17:19           [0, 140]   
1542           NaN          NaN 2016-01-19 18:39:13           [0, 113]   
1904           NaN          NaN 2015-12-09 06:01:26            [0, 91]   
715            NaN          NaN 2016-10-06 00:08:09            [0, 77]   
2283           NaN          NaN 2015-11-19 03:52:34           [0, 136]   

                                               entities  \
313   {'hashtags': [], 'symbols': [], 'user_mentions...   
1266  {'hashtags': [], 'symbols': [], 'user_mentions...   
1942  {'hashtags': [], 'symbols': [], 'user_mentions...   
2109  {'hashtags': [], 'symbols': [], 'user_mentions...   
2028  {'hashtags': [], 'symbols': [], 'user_mentions...   
385   {'hashtags': [], 'symbols': [], 'user_mentions...   
1542  {'hashtags': [], 'symbols': [], 'user_mentions...   
1904  {'hashtags': [], 'symbols': [], 'user_mentions...   
715   {'hashtags': [], 'symbols': [], 'user_mentions...   
2283  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  favorite_count  \
313   {'media': [{'id': 835172763019194370, 'id_str'...           28552   
1266  {'media': [{'id': 709556945439617027, 'id_str'...            3593   
1942  {'media': [{'id': 673708601312178176, 'id_str'...            1153   
2109  {'media': [{'id': 670442333423452160, 'id_str'...             690   
2028  {'media': [{'id': 671789647081644032, 'id_str'...            7527   
385                                                 NaN               0   
1542  {'media': [{'id': 689517477928292354, 'id_str'...            3735   
1904  {'media': [{'id': 674468873001938944, 'id_str'...            6676   
715   {'media': [{'id': 783821100060905476, 'id_str'...            8209   
2283  {'media': [{'id': 667188686895861760, 'id_str'...             784   

      favorited                                          full_text  geo  \
313       False  We only rate dogs. Please don't send in any no...  NaN   
1266      False  This is Cecil. She's a Gigglefloof Poofer. Out...  NaN   
1942      False  This is Riley. She's just an adorable football...  NaN   
2109      False  Meet Koda. He's large. Looks very soft. Great ...  NaN   
2028      False  This is space pup. He's very confused. Tries t...  NaN   
385       False  RT @dog_rates: Please only send in dogs. We on...  NaN   
1542      False  This is Carl. He just wants to make sure you'r...  NaN   
1904      False  This is Louis. He thinks he's flying. 13/10 th...  NaN   
715       False  This is Mairi. She has mastered the art of cam...  NaN   
2283      False  Quite an advanced dog here. Impressively dress...  NaN   

                            ...                          \
313                         ...                           
1266                        ...                           
1942                        ...                           
2109                        ...                           
2028                        ...                           
385                         ...                           
1542                        ...                           
1904                        ...                           
715                         ...                           
2283                        ...                           

      possibly_sensitive_appealable  quoted_status quoted_status_id  \
313                             0.0            NaN              NaN   
1266                            0.0            NaN              NaN   
1942                      

## Programatic Assessment Part

In [49]:
# check for duplicates
assert [twitter_archive.duplicated() == False]
assert [image_predictions.duplicated() == False]
assert [tweet_status.id.duplicated() == False]

There are no duplicates 

In [50]:
# checking numbers of dog breeds
image_predictions.p1.value_counts()

golden_retriever             150
Labrador_retriever           100
Pembroke                      89
Chihuahua                     83
pug                           57
chow                          44
Samoyed                       43
toy_poodle                    39
Pomeranian                    38
cocker_spaniel                30
malamute                      30
French_bulldog                26
Chesapeake_Bay_retriever      23
miniature_pinscher            23
seat_belt                     22
Siberian_husky                20
Staffordshire_bullterrier     20
German_shepherd               20
Cardigan                      19
web_site                      19
Eskimo_dog                    18
beagle                        18
teddy                         18
Maltese_dog                   18
Shetland_sheepdog             18
Shih-Tzu                      17
Rottweiler                    17
Lakeland_terrier              17
Italian_greyhound             16
kuvasz                        16
          

In [52]:
# excluding Less Likely to Appear (LLTA) breads 
LLTA = image_predictions.groupby('p1').filter(lambda x: len(x) < 5)

In [53]:
LLTA.head()

tweet_id                                          jpg_url  \
0   666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
3   666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
6   666051853826850816  https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
11  666071193221509120  https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
17  666104133288665088  https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   

    img_num                      p1   p1_conf  p1_dog                 p2  \
0         1  Welsh_springer_spaniel  0.465074    True             collie   
3         1     Rhodesian_ridgeback  0.408143    True            redbone   
6         1              box_turtle  0.933012   False         mud_turtle   
11        1           Gordon_setter  0.503672    True  Yorkshire_terrier   
17        1                     hen  0.965932   False               cock   

     p2_conf  p2_dog                  p3   p3_conf  p3_dog  
0   0.156665    True   Shetland_sheepdog  0.061428    True  
3   0.360687    True  miniature_pinscher  0.222752    True  
6   0.045885   False            terrapin  0.017885   False  
11  0.174201    True            Pekinese  0.109454    True  
17  0.033919   False           partridge  0.000052   False

In [57]:
len(LLTA)

473

In [58]:
len(image_predictions)

2075

In [59]:
tweet_status

contributors  coordinates          created_at display_text_range  \
0              NaN          NaN 2017-08-01 16:23:56            [0, 85]   
1              NaN          NaN 2017-08-01 00:17:27           [0, 138]   
2              NaN          NaN 2017-07-31 00:18:03           [0, 121]   
3              NaN          NaN 2017-07-30 15:58:51            [0, 79]   
4              NaN          NaN 2017-07-29 16:00:24           [0, 138]   
5              NaN          NaN 2017-07-29 00:08:17           [0, 138]   
6              NaN          NaN 2017-07-28 16:27:12           [0, 140]   
7              NaN          NaN 2017-07-28 00:22:40           [0, 118]   
8              NaN          NaN 2017-07-27 16:25:51           [0, 122]   
9              NaN          NaN 2017-07-26 15:59:51           [0, 133]   
10             NaN          NaN 2017-07-26 00:31:25           [0, 130]   
11             NaN          NaN 2017-07-25 16:11:53           [0, 107]   
12             NaN          NaN 2017-07-25 01:55:32           [0, 106]   
13             NaN          NaN 2017-07-25 00:10:02            [0, 91]   
14             NaN          NaN 2017-07-24 17:02:04           [0, 118]   
15             NaN          NaN 2017-07-24 00:19:32           [0, 138]   
16             NaN          NaN 2017-07-23 00:22:39            [0, 86]   
17             NaN          NaN 2017-07-22 16:56:37           [0, 128]   
18             NaN          NaN 2017-07-22 00:23:06            [0, 87]   
19             NaN          NaN 2017-07-20 16:49:33           [0, 127]   
20             NaN          NaN 2017-07-19 16:06:48           [0, 127]   
21             NaN          NaN 2017-07-19 03:39:09           [0, 108]   
22             NaN          NaN 2017-07-19 00:47:34            [0, 99]   
23             NaN          NaN 2017-07-18 16:08:03            [0, 88]   
24             NaN          NaN 2017-07-18 00:07:08           [0, 129]   
25             NaN          NaN 2017-07-17 16:17:36           [0, 101]   
26             NaN          NaN 2017-07-16 23:58:41           [0, 121]   
27             NaN          NaN 2017-07-16 20:14:00            [0, 71]   
28             NaN          NaN 2017-07-15 23:25:31           [0, 131]   
29             NaN          NaN 2017-07-15 16:51:35          [27, 105]   
...            ...          ...                 ...                ...   
2324           NaN          NaN 2015-11-17 00:24:19           [0, 140]   
2325           NaN          NaN 2015-11-17 00:06:54           [0, 139]   
2326           NaN          NaN 2015-11-16 23:23:41           [0, 137]   
2327           NaN          NaN 2015-11-16 21:54:18            [0, 81]   
2328           NaN          NaN 2015-11-16 21:10:36           [0, 140]   
2329           NaN          NaN 2015-11-16 20:32:58           [0, 135]   
2330           NaN          NaN 2015-11-16 20:01:42           [0, 112]   
2331           NaN          NaN 2015-11-16 19:31:45           [0, 139]   
2332           NaN          NaN 2015-11-16 16:37:02           [0, 138]   
2333           NaN          NaN 2015-11-16 16:11:11           [0, 136]   
2334           NaN          NaN 2015-11-16 15:14:19            [0, 46]   
2335           NaN          NaN 2015-11-16 14:57:41            [0, 82]   
2336           NaN          NaN 2015-11-16 04:02:55           [0, 134]   
2337           NaN          NaN 2015-11-16 03:55:04           [0, 128]   
2338           NaN          NaN 2015-11-16 03:44:34           [0, 140]   
2339           NaN          NaN 2015-11-16 03:22:39           [0, 132]   
2340           NaN          NaN 2015-11-16 02:38:37           [0, 125]   
2341           NaN          NaN 2015-11-16 01:59:36           [0, 137]   
2342           NaN          NaN 2015-11-16 01:52:02           [0, 137]   
2343           NaN          NaN 2015-11-16 01:22:45           [0, 107]   
2344           NaN          NaN 2015-11-16 01:01:59           [0, 135]   
2345           NaN          NaN 2015-11-16 00:55:59           [0, 124]   
2346      

In [60]:
tweet_status.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2354 non-null datetime64[ns]
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null float64
id                               2354 non-null int64
id_str                           2354 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

In [61]:
twitter_archive.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [62]:
# comparing tweet_status.created_at and twitter_archive.timestamp

df1 = tweet_status.filter(['id','created_at'])
df1 = df1.rename(columns={'id': 'tweet_id'})

df2 = tw_archive.filter(['tweet_id', 'timestamp'])

df_merged = pd.merge(df1, df2, how='inner', on=['tweet_id'])

In [63]:
df_merged.head()

tweet_id          created_at                  timestamp
0  892420643555336193 2017-08-01 16:23:56  2017-08-01 16:23:56 +0000
1  892177421306343426 2017-08-01 00:17:27  2017-08-01 00:17:27 +0000
2  891815181378084864 2017-07-31 00:18:03  2017-07-31 00:18:03 +0000
3  891689557279858688 2017-07-30 15:58:51  2017-07-30 15:58:51 +0000
4  891327558926688256 2017-07-29 16:00:24  2017-07-29 16:00:24 +0000

# Assessment Reporting

## Quality issues

### Twitter_archive

- Too many values and might not be needed in this analysis.

- Delete the cells with numerator exceeding 20.
- For the column “rating_numirator” there are zero values. 
- For the column “rating_numirator” there are outlier values.

- For the column “rating_denumirator” there is a zero value.
- For the column “rating_denumirator” there are extreme values.

- Some columns like (in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id) are missing.
- Typo mistakes in the nomirator and the denominator.

### image_predictions

- There are unneeded columns
- Cleaning irrelevant rows like “Egyptian_cat”.
- Cleaning irrelevant values in p1 like “school_bus, pillow, cartoon”.
- 473 rows (frequency less than 5) because they are less frequant and most likely not dogs.

### tweet_status

- unneeded columns

## Tidiness Issues 

### Twitter_archive

- unnessesarry columns 
- Replacing 'None' with np.nan to indicate the missing values
- contain one column for dog classification instead of several columns



### image_predictions

-   not descriptive names (p1, p2 and p3). 


### tweet_status

- more convinient naming needed(id to be tweet_id) to ease merging
- tweet_status.created_at is the same as the column tw_archive.timestamp

# Cleaning the Data

## Cleaning archive_clean

In [37]:
# first, creating new data frame from the original
archive_clean = twitter_archive.copy()

In [38]:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [39]:
# define: dropping all retweets

# code:
archive_clean = archive_clean[archive_clean['retweeted_status_id'].isnull()]
archive_clean = archive_clean[archive_clean['retweeted_status_user_id'].isnull()]
archive_clean = archive_clean[archive_clean['retweeted_status_timestamp'].isnull()]


In [40]:
# test:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2175 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2175 non-null object
source                        2175 non-null object
text                          2175 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2117 non-null object
rating_numerator              2175 non-null int64
rating_denominator            2175 non-null int64
name                          2175 non-null object
doggo                         2175 non-null object
floofer                       2175 non-null object
pupper                        2175 non-null object
puppo                         2175 non-null object
dtypes: float64(4), int64(3), object(1

In [41]:
#test: no text containing (RT) 
archive_clean

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
30    886267009285017600           8.862664e+17         2.281182e+09   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [42]:
# define: dropping all unnessessary columns that miss so many values and unneeded ones like expanded_urls

# code:
archive_clean = archive_clean.drop(['in_reply_to_status_id', 
                                        'in_reply_to_user_id',
                                        'retweeted_status_id',
                                        'retweeted_status_user_id',
                                        'retweeted_status_timestamp',
                                         'expanded_urls'], axis=1)

In [43]:
# test:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2175 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2175 non-null int64
timestamp             2175 non-null object
source                2175 non-null object
text                  2175 non-null object
rating_numerator      2175 non-null int64
rating_denominator    2175 non-null int64
name                  2175 non-null object
doggo                 2175 non-null object
floofer               2175 non-null object
pupper                2175 non-null object
puppo                 2175 non-null object
dtypes: int64(3), object(8)
memory usage: 203.9+ KB


In [44]:
#define: fixing typo mistakes

#code:
archive_clean.loc[archive_clean['rating_numerator']==80, ['rating_numerator']] = 10
archive_clean.loc[archive_clean['rating_denominator']==80, ['rating_denominator']] = 10
archive_clean.loc[archive_clean['rating_numerator']==44, ['rating_numerator']] = 11
archive_clean.loc[archive_clean['rating_denominator']==40, ['rating_denominator']] = 10


In [45]:
#test:
len(archive_clean[archive_clean['rating_numerator'] > 15])

23

In [46]:
#define: cleaning irrelevant data like zeros or outliers

#code:
archive_clean = archive_clean[archive_clean['rating_numerator'] != 0 ]
archive_clean = archive_clean[archive_clean['rating_denominator'] >= 10 ]
archive_clean = archive_clean[archive_clean['rating_numerator'] <= 20 ]
archive_clean = archive_clean[archive_clean['rating_denominator'] < 20 ]

In [47]:
#test:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2149 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2149 non-null int64
timestamp             2149 non-null object
source                2149 non-null object
text                  2149 non-null object
rating_numerator      2149 non-null int64
rating_denominator    2149 non-null int64
name                  2149 non-null object
doggo                 2149 non-null object
floofer               2149 non-null object
pupper                2149 non-null object
puppo                 2149 non-null object
dtypes: int64(3), object(8)
memory usage: 201.5+ KB


In [48]:
# test:
len(archive_clean[archive_clean['rating_numerator'] == 0])

0

In [49]:
# test:
len(archive_clean[archive_clean['rating_numerator'] > 20])

0

In [50]:
# test:
len(archive_clean[archive_clean['rating_denominator'] < 10])

0

In [51]:
# test:
len(archive_clean[archive_clean['rating_denominator'] > 20])

0

In [52]:
# test:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2149 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2149 non-null int64
timestamp             2149 non-null object
source                2149 non-null object
text                  2149 non-null object
rating_numerator      2149 non-null int64
rating_denominator    2149 non-null int64
name                  2149 non-null object
doggo                 2149 non-null object
floofer               2149 non-null object
pupper                2149 non-null object
puppo                 2149 non-null object
dtypes: int64(3), object(8)
memory usage: 201.5+ KB


In [53]:
# define: cleaning None values with NaN

#code:
archive_clean =  archive_clean.replace('None', np.nan)

In [54]:
# test:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2149 entries, 0 to 2355
Data columns (total 11 columns):
tweet_id              2149 non-null int64
timestamp             2149 non-null object
source                2149 non-null object
text                  2149 non-null object
rating_numerator      2149 non-null int64
rating_denominator    2149 non-null int64
name                  1487 non-null object
doggo                 87 non-null object
floofer               10 non-null object
pupper                233 non-null object
puppo                 25 non-null object
dtypes: int64(3), object(8)
memory usage: 201.5+ KB


In [55]:
#define: creating new column to combine dog classifcations

#code:
archive_clean.loc[archive_clean['doggo'] == 'doggo', 'dog_class'] = 'doggo'
archive_clean.loc[archive_clean['floofer'] == 'floofer', 'dog_class'] = 'floofer'
archive_clean.loc[archive_clean['pupper'] == 'pupper', 'dog_class'] = 'pupper'
archive_clean.loc[archive_clean['puppo'] == 'puppo', 'dog_class'] = 'puppo'

In [56]:
#test:
archive_clean.head(10)

tweet_id                  timestamp  \
0  892420643555336193  2017-08-01 16:23:56 +0000   
1  892177421306343426  2017-08-01 00:17:27 +0000   
2  891815181378084864  2017-07-31 00:18:03 +0000   
3  891689557279858688  2017-07-30 15:58:51 +0000   
4  891327558926688256  2017-07-29 16:00:24 +0000   
5  891087950875897856  2017-07-29 00:08:17 +0000   
6  890971913173991426  2017-07-28 16:27:12 +0000   
7  890729181411237888  2017-07-28 00:22:40 +0000   
8  890609185150312448  2017-07-27 16:25:51 +0000   
9  890240255349198849  2017-07-26 15:59:51 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   
5  <a href="http://twitter.com/download/iphone" r...   
6  <a href="http://twitter.com/download/iphone" r...   
7  <a href="http://twitter.com/download/iphone" r...   
8  <a href="http://twitter.com/download/iphone" r...   
9  <a href="http://twitter.com/download/iphone" r...   

                                                text  rating_numerator  \
0  This is Phineas. He's a mystical boy. Only eve...                13   
1  This is Tilly. She's just checking pup on you....                13   
2  This is Archie. He is a rare Norwegian Pouncin...                12   
3  This is Darla. She commenced a snooze mid meal...                13   
4  This is Franklin. He would like you to stop ca...                12   
5  Here we have a majestic great white breaching ...                13   
6  Meet Jax. He enjoys ice cream so much he gets ...                13   
7  When you watch your owner call another dog a g...                13   
8  This is Zoey. She doesn't want to be one of th...                13   
9  This is Cassie. She is a college pup. Studying...                14   

   rating_denominator      name  doggo floofer pupper puppo dog_class  
0                  10   Phineas    NaN     NaN    NaN   NaN       NaN  
1                  10     Tilly    NaN     NaN    NaN   NaN       NaN  
2                  10    Archie    NaN     NaN    NaN   NaN       NaN  
3                  10     Darla    NaN     NaN    NaN   NaN       NaN  
4                  10  Franklin    NaN     NaN    NaN   NaN       NaN  
5                  10       NaN    NaN     NaN    NaN   NaN       NaN  
6                  10       Jax    NaN     NaN    NaN   NaN       NaN  
7                  10       NaN    NaN     NaN    NaN   NaN       NaN  
8                  10      Zoey    NaN     NaN    NaN   NaN       NaN  
9                  10    Cassie  doggo     NaN    NaN   NaN     doggo

In [57]:
# define: droping the redundent columns 

# code:
archive_clean = archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 1)

In [58]:
# test:
archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2149 entries, 0 to 2355
Data columns (total 8 columns):
tweet_id              2149 non-null int64
timestamp             2149 non-null object
source                2149 non-null object
text                  2149 non-null object
rating_numerator      2149 non-null int64
rating_denominator    2149 non-null int64
name                  1487 non-null object
dog_class             343 non-null object
dtypes: int64(3), object(5)
memory usage: 151.1+ KB


## Cleaning image.predictions

In [59]:
images_clean = image_predictions.copy()

In [60]:
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [62]:
# define: renaming columns to be descriptive

# code:
images_clean= images_clean.rename(columns={'p1':'BreedFirstProbability',
                                               'p2':'BreedSecondProbability',
                                               'p3':'BreedThirdProbability'})


In [63]:
# test
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id                  2075 non-null int64
jpg_url                   2075 non-null object
img_num                   2075 non-null int64
BreedFirstProbability     2075 non-null object
p1_conf                   2075 non-null float64
p1_dog                    2075 non-null bool
BreedSecondProbability    2075 non-null object
p2_conf                   2075 non-null float64
p2_dog                    2075 non-null bool
BreedThirdProbability     2075 non-null object
p3_conf                   2075 non-null float64
p3_dog                    2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [64]:
# define: dropping all probabilities less than the half

# code:
images_clean = images_clean[images_clean['p1_conf'] >= 0.5]

In [65]:
#test: 
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1239 entries, 1 to 2072
Data columns (total 12 columns):
tweet_id                  1239 non-null int64
jpg_url                   1239 non-null object
img_num                   1239 non-null int64
BreedFirstProbability     1239 non-null object
p1_conf                   1239 non-null float64
p1_dog                    1239 non-null bool
BreedSecondProbability    1239 non-null object
p2_conf                   1239 non-null float64
p2_dog                    1239 non-null bool
BreedThirdProbability     1239 non-null object
p3_conf                   1239 non-null float64
p3_dog                    1239 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 100.4+ KB


In [66]:
# define: selecting only the three needed columns to work with

# code:
images_clean = images_clean.filter(['tweet_id',
                                    'BreedFirstProbability',
                                    'p1_conf'] )


In [67]:
#test
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1239 entries, 1 to 2072
Data columns (total 3 columns):
tweet_id                 1239 non-null int64
BreedFirstProbability    1239 non-null object
p1_conf                  1239 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 38.7+ KB


In [68]:
# define: dropping the less frequent stuff in the data

#code
images_clean = images_clean.groupby('BreedFirstProbability').filter(lambda x: len(x) > 5)

In [69]:
#test
images_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 881 entries, 2 to 2072
Data columns (total 3 columns):
tweet_id                 881 non-null int64
BreedFirstProbability    881 non-null object
p1_conf                  881 non-null float64
dtypes: float64(1), int64(1), object(1)
memory usage: 27.5+ KB


## Cleaning tweet_status

In [70]:
status_clean = tweet_status.copy()

In [71]:
status_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2354 non-null datetime64[ns]
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null float64
id                               2354 non-null int64
id_str                           2354 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

In [72]:
# define: changing id to tweet_id to ease combining

#code
status_clean = status_clean.rename(columns={'id':'tweet_id'})

In [73]:
#test
status_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 31 columns):
contributors                     0 non-null float64
coordinates                      0 non-null float64
created_at                       2354 non-null datetime64[ns]
display_text_range               2354 non-null object
entities                         2354 non-null object
extended_entities                2073 non-null object
favorite_count                   2354 non-null int64
favorited                        2354 non-null bool
full_text                        2354 non-null object
geo                              0 non-null float64
tweet_id                         2354 non-null int64
id_str                           2354 non-null int64
in_reply_to_screen_name          78 non-null object
in_reply_to_status_id            78 non-null float64
in_reply_to_status_id_str        78 non-null float64
in_reply_to_user_id              78 non-null float64
in_reply_to_user_id_str          78 n

In [74]:
#define: selecting only the needed columns 

#code:
status_clean = status_clean.filter(['tweet_id',
                                    'favorite_count',
                                    'retweet_count',
                                    'source',
                                    'user'])

In [75]:
#test:
status_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 5 columns):
tweet_id          2354 non-null int64
favorite_count    2354 non-null int64
retweet_count     2354 non-null int64
source            2354 non-null object
user              2354 non-null object
dtypes: int64(3), object(2)
memory usage: 92.0+ KB


## Merging all three data frames with tweet_id

In [76]:
final_df = pd.merge(archive_clean, images_clean, how='outer', on=['tweet_id'])

In [77]:
final_df = pd.merge(final_df, status_clean, how='outer', on=['tweet_id'])

In [78]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2355 entries, 0 to 2354
Data columns (total 14 columns):
tweet_id                 2355 non-null int64
timestamp                2149 non-null object
source_x                 2149 non-null object
text                     2149 non-null object
rating_numerator         2149 non-null float64
rating_denominator       2149 non-null float64
name                     1487 non-null object
dog_class                343 non-null object
BreedFirstProbability    881 non-null object
p1_conf                  881 non-null float64
favorite_count           2354 non-null float64
retweet_count            2354 non-null float64
source_y                 2354 non-null object
user                     2354 non-null object
dtypes: float64(5), int64(1), object(8)
memory usage: 276.0+ KB


In [79]:
final_df

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
5     891087950875897856  2017-07-29 00:08:17 +0000   
6     890971913173991426  2017-07-28 16:27:12 +0000   
7     890729181411237888  2017-07-28 00:22:40 +0000   
8     890609185150312448  2017-07-27 16:25:51 +0000   
9     890240255349198849  2017-07-26 15:59:51 +0000   
10    890006608113172480  2017-07-26 00:31:25 +0000   
11    889880896479866881  2017-07-25 16:11:53 +0000   
12    889665388333682689  2017-07-25 01:55:32 +0000   
13    889638837579907072  2017-07-25 00:10:02 +0000   
14    889531135344209921  2017-07-24 17:02:04 +0000   
15    889278841981685760  2017-07-24 00:19:32 +0000   
16    888917238123831296  2017-07-23 00:22:39 +0000   
17    888804989199671297  2017-07-22 16:56:37 +0000   
18    888554962724278272  2017-07-22 00:23:06 +0000   
19    888078434458587136  2017-07-20 16:49:33 +0000   
20    887705289381826560  2017-07-19 16:06:48 +0000   
21    887517139158093824  2017-07-19 03:39:09 +0000   
22    887473957103951883  2017-07-19 00:47:34 +0000   
23    887343217045368832  2017-07-18 16:08:03 +0000   
24    887101392804085760  2017-07-18 00:07:08 +0000   
25    886983233522544640  2017-07-17 16:17:36 +0000   
26    886736880519319552  2017-07-16 23:58:41 +0000   
27    886680336477933568  2017-07-16 20:14:00 +0000   
28    886366144734445568  2017-07-15 23:25:31 +0000   
29    886267009285017600  2017-07-15 16:51:35 +0000   
...                  ...                        ...   
2325  768554158521745409                        NaN   
2326  766864461642756096                        NaN   
2327  766078092750233600                        NaN   
2328  763167063695355904                        NaN   
2329  761371037149827077                        NaN   
2330  760153949710192640                        NaN   
2331  759159934323924993                        NaN   
2332  758467244762497024                        NaN   
2333  757729163776290825                        NaN   
2334  757597904299253760                        NaN   
2335  754874841593970688                        NaN   
2336  753298634498793472                        NaN   
2337  752701944171524096                        NaN   
2338  752309394570878976                        NaN   
2339  749981277374128128                        NaN   
2340  747242308580548608                        NaN   
2341  746906459439529985                        NaN   
2342  746521445350707200                        NaN   
2343  743835915802583040                        NaN   
2344  731156023742988288                        NaN   
2345  722974582966214656                        NaN   
2346  716439118184652801                        NaN   
2347  713900603437621249                        NaN   
2348  711998809858043904                        NaN   
2349  709198395643068416                        NaN   
2350  686035780142297088                        NaN   
2351  684225744407494656                        NaN   
2352  684222868335505415                        NaN   
2353  680494726643068929                        NaN   
2354  670842764863651840                        NaN   

                                               source_x  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
5     <a href="http://twitter.com/download/iphone" r...   
6     <a href="http://twitter.com/download/iphone" r...   
7     <a href="http://twitter.com/download/iphone" r...   
8     <a href="http://twitter.com/download/iphone" r...   
9     <a href="

# Saving as CSV

In [81]:
final_df.to_csv('twitter_archive_master.csv')